contracts/core/space.cairo

```diff
@@ -450,35 +442,144 @@ func _handle_collision_with_other_ship{
+    let (contract_address) = get_contract_address()
+    dust_destroyed.emit(contract_address, position)
 
     return ()
 end
 
 func _burn_dust{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
-        dust_id : Uint256):
-    let (space_contract_address) = get_contract_address()
-    let (dust_contract_address) = dust_contract.read()
+        dust : Dust, position : Vector2):
+    assert dust.present = TRUE
 
-    let (dust) = IDustContract.metadata(dust_contract_address, dust_id)
-    IDustContract.burn(dust_contract_address, dust_id)
     let (dust_count) = current_dust_count.read()
     assert_nn(dust_count)
     current_dust_count.write(dust_count - 1)
 
-    dust_destroyed.emit(space_contract_address, dust_id, dust.position)
+    let (contract_address) = get_contract_address()
+    dust_destroyed.emit(contract_address, position)
 
     return ()
 end
+
+# Generate random dust given a space size
+func _generate_random_dust_on_border{
+        pedersen_ptr : HashBuiltin*, syscall_ptr : felt*, range_check_ptr,
+        bitwise_ptr : BitwiseBuiltin*}() -> (dust : Dust, position : Vector2):
+    alloc_locals
+    local dust : Dust
+    assert dust.present = TRUE
+
+    let (seed) = dust_generation_seed.read()
+    dust_generation_seed.write(seed + 1)
+
+    let (rand_contract_address) = rand_contract.read()
+    let (r1, r2, r3, r4, r5) = IRandom.generate_random_numbers(
+        rand_contract_address, seed)
+
+    let (direction : Vector2) = MathUtils_random_direction(r1, r2)
+    assert dust.direction = direction
+
+    let (position : Vector2) = _generate_random_position_on_border(r3, r4, r5)
+
+    return (dust=dust, position=position)
+end
+
+# Generate a random position on a given border (top, left, right, bottom)
+func _generate_random_position_on_border{
+        pedersen_ptr : HashBuiltin*, syscall_ptr : felt*, range_check_ptr}(r1, r2, r3) -> (position : Vector2):
+    alloc_locals
+    let (space_size) = _get_grid_size()
+
+    # x is 0 or space_size - 1
+    let (x) = MathUtils_random_in_range(r1, 0, 1)
+    local x = x * (space_size - 1)
+
+    # y is in [0, space_size-1]
+    let (y) = MathUtils_random_in_range(r2, 0, space_size - 1)
+
+    return _shuffled_position(x, y, r3)
+end
+
+# given x, y return randomly Position(x,y) or Position(y,x)
+func _shuffled_position{pedersen_ptr : HashBuiltin*, syscall_ptr : felt*, range_check_ptr}(
+        x : felt, y : felt, r) -> (position : Vector2):
+    alloc_locals
+    local position : Vector2
+
+    let (on_horizontal_border) = MathUtils_random_in_range(r, 0, 1)
+    if on_horizontal_border == 0:
+        assert position.x = x
+        assert position.y = y
+    else:
+        assert position.x = y
+        assert position.y = x
+    end
+
+    return (position=position)
+end
+
+
+func _compute_new_dust_position{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(dust : Dust, current_position : Vector2) -> (
```

Comment:
This is not very clear to me what this method does based on its name. Is it a random new direction?

Reply:
it computes the new position of the dust, based on its current position and direction, and handles bounces on borders at the same time.